prereq-build: fix inconsistent value of $PATH

mcprat opened the pull request:
refer to
#12617 
#12700

mcprat committed:
treewide: add ORIG_PATH variable

Add a variable that stores the original value of $PATH
in the host system's shell, before Make alters it.

This can be useful for when it is necessary
to ignore symlinks and programs made by the build system.

Define this new variable before all instances of
'export PATH:=' or similar.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
prereq-build: fix inconsistent value of $PATH

In the recipe SetupHostCommand for checking
and creating symlinks, $PATH was only overridden
for one of several commands.

This causes the symlinks to be included
in the paths to pick a program from
when the check was repeated, because
staging_dir/host/bin was included in $PATH,
but only *sometimes*.

When the check ran again, the command succeded
with a $PATH including the symlink,
(eval "$$$$$$$$cmd")
while the path to the program was evaluated
with a $PATH NOT including the symlink,
(bin=...)
causing the symlink to be relinked incorrectly,
instead of passing as exactly the same.

Coincidentally, this was only a problem
if the symlink targeted the alternative
program with a different name.

By overriding the value of $PATH at the invocation of Make,
it will apply to the entire environment of the checks.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
Revert "prereq: SetupHostCommand: fix wrong check result"

This reverts commit 7855378fcd7ed7cb0a223238a99bac0b8e46c380.

The return "exit 1" was intentional and actually just
makes the symlink checks much more strict.

This new level of strictness added to the checks revealed
what was a confusing regression with prereq stage that
already existed but was not presenting itself
because of the simple way that checks used to be done before.

Either way, reverting to "exit 0" was a nice workaround
until the true root cause was discovered, so as to not interfere
with others' pull requests and builds in the meantime.

It turns out that this problem was the inconsistent value of $PATH
between different commands within the SetupHostCommand recipe,
now fixed in the parent commit, using the variable created
in the parent of the parent commit.

Ref: f75204036c ("prereq-build: allow host command symlinks to update")
Signed-off-by: Michael Pratt <mcpratt@pm.me>

dhewg commented on target/toolchain/files/wrapper.sh:
```diff
@@ -43,6 +43,7 @@ TOOLCHAIN_BIN_DIR="$REALNAME_DIR/"
 
 # Set the PATH so that our run-time location is first
 # (get_feature is run from the path, so this has to be set)
+export ORIG_PATH=${ORIG_PATH:-$PATH}
```
This looks like it may be redundant? It's just a script that does $stuff without calling $MAKE again?

mpratt14 replied:
first commit, I'm just putting a variable that stores the original $PATH in every place where it looks like $PATH is changing.

it may not be useful there, but eh why not?

mpratt14 replied:
for consistency at least...

dhewg commented:
Do you really need to spam `export ORIG_PATH:=...` everywhere `PATH` gets modified?

Shouldn't this in toplevel.mk be sufficient?
```
ifeq ($(origin ORIG_PATH),undefined)
export ORIG_PATH:=$(PATH)
endif
```

mpratt14 commented:
well at least in every Makefile because PATH will be modified by the time toplevel.mk is included

ynezz commented:
Nice, thanks! @ldir-EDB0 could you test drive this?

> Do you really need to spam export ORIG_PATH:=... everywhere PATH gets modified?

Well, I don't like it as well, but it should be rock solid solution compared to some manual filtering, and reusable as well.

ldir-EDB0 commented:
It works in that the install pre-req is found, no problem.
It doesn't work in that after a build has gone through another build will update at least 4 links in $STAGING_DIR/host/bin and it seems include a version of something that means patches no longer apply.

So an initial build works, subsequent builds fail.

mpratt14 commented:
OK, so can you share which links change and the error you get?

mpratt14 commented:
@ldir-EDB0 but the links are changing on a second build, as well as the first?

do you do `make clean` in between or no?

mpratt14 commented:
> So an initial build works, subsequent builds fail.

to be honest this doesn't make sense to me

prereq-build has a stampfile. If you build once, symlinks update, if you build again, prereq-build symlink checks are not run at all, unless you did a `make dirclean`, but if you did that would be equivalent to an "initial build". Or, are you just deleting that stampfile alone or re-edit one of the dependent files?

I can't reproduce any issue, please show us what you're seeing @ldir-EDB0

ldir-EDB0 commented:
I'm on work shift until at least monday so can't fit in openwrt issues.  Also, to be honest, I don't know how to give you what you want.  I have ls -l of the host/bin dir for the many different attempts at building but I don't know how to condense that into something useful.  Also my build machine is very slow.  There's a definite problem though. the sequence I run through, make dirclean, make -j4, then another make -j4 will update the links again and then fail attempting to apply kernel patches.  I think the build puts in 'local' versions of some utilities, incl say 'bash' which are newer than prereq-build check, so it runs again and replaces them with the 'native' pre-req links rather than leeping the openwrt built versions.  That's why I see the issue after 1st build.

mpratt14 commented:
take your time, no problem

start with
`ls -l staging_dir/host/bin | grep '\->'`
after first build and then after second build to compare the two cases

and the actual error that stops the second build (last 20 or so lines of output with `V=sc`)

mcprat committed:
prereq-build: do not replace binaries with symlinks

Some programs, like bash and patch, are checked by prereq stage
and have a symlink installed, but then is later built from source.

Now that the prereq-build checks are not successful
just by finding the file alone, it is possible for
a new symlink to overwrite the installed binary.

If a normal file is found in staging_dir/host/bin,
let the check look for the associated stampfile, and if found,
skip creation of a symlink and exit successfully.

Suggested-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I looked into the case of bash, and I see the problem... I added a commit to fix that

I don't see how which version of bash has anything to do with applying kernel patches, but maybe that was it.

mpratt14 commented:
so the same case for bash probably happens for patch too, because there is a SetupHostCommand for patch as well

mpratt14 commented:
> another make -j4 will update the links again

I feel like this should not happen though

if I delete some symlinks and run `make prereq` on my system, the checks don't get run again...

ynezz commented:
> if I delete some symlinks and run make prereq on my system, the checks don't get run again...

I've tried to [reproduce the issue](https://github.com/ynezz/openwrt/actions/runs/5090605078/jobs/9149635380) but it seems to be broken currently (probably not related to this issue, though):

<details>

<summary>Make errors from logs/package/libs/libxml2/host-compile.txt</summary>

```shell
logs/package/libs/libxml2/host-compile.txt--- Performing Test Iconv_IS_BUILT_IN - Failed
logs/package/libs/libxml2/host-compile.txt-CMake Error at /Volumes/OpenWrt/openwrt/staging_dir/host/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
logs/package/libs/libxml2/host-compile.txt-  Could NOT find Iconv (missing: Iconv_LIBRARY Iconv_INCLUDE_DIR)
logs/package/libs/libxml2/host-compile.txt-Call Stack (most recent call first):
logs/package/libs/libxml2/host-compile.txt-  /Volumes/OpenWrt/openwrt/staging_dir/host/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
logs/package/libs/libxml2/host-compile.txt-  /Volumes/OpenWrt/openwrt/staging_dir/host/share/cmake-3.26/Modules/FindIconv.cmake:165 (find_package_handle_standard_args)
logs/package/libs/libxml2/host-compile.txt-  CMakeLists.txt:67 (find_package)
logs/package/libs/libxml2/host-compile.txt-
logs/package/libs/libxml2/host-compile.txt-
logs/package/libs/libxml2/host-compile.txt--- Configuring incomplete, errors occurred!
logs/package/libs/libxml2/host-compile.txt:make[3]: *** [Makefile:205: /Volumes/OpenWrt/openwrt/build_dir/hostpkg/libxml2-2.11.4/.configured] Error 1
logs/package/libs/libxml2/host-compile.txt-time: package/libs/libxml2/host-compile#1.13#1.65#41.00
```

</details>

<details>
<summary>Make errors from logs/tools/libtool/compile.txt</summary>

```shell
logs/tools/libtool/compile.txt-
logs/tools/libtool/compile.txt-Applying ./patches/140-don-t-quote-SHELL-in-Makefile.am.patch using plaintext: 
logs/tools/libtool/compile.txt-patching file Makefile.am
logs/tools/libtool/compile.txt-
logs/tools/libtool/compile.txt-Applying ./patches/200-openwrt-branding.patch using plaintext: 
logs/tools/libtool/compile.txt-patching file build-aux/ltmain.in
logs/tools/libtool/compile.txt-patching file build-aux/funclib.sh
logs/tools/libtool/compile.txt-make[4]: Entering directory '/Volumes/OpenWrt/openwrt/build_dir/host/libtool-2.4.7'
logs/tools/libtool/compile.txt-There seems to be no Makefile in this directory.
logs/tools/libtool/compile.txt-You must run ./configure before running 'make'.
logs/tools/libtool/compile.txt:make[4]: *** [GNUmakefile:108: abort-due-to-no-makefile] Error 1
logs/tools/libtool/compile.txt-make[4]: Leaving directory '/Volumes/OpenWrt/openwrt/build_dir/host/libtool-2.4.7'
logs/tools/libtool/compile.txt-bootstrap: warning: No 'git' found; imported gnulib modules may be outdated.
logs/tools/libtool/compile.txt-bootstrap: running: git submodule init -- gl-mod/bootstrap
logs/tools/libtool/compile.txt-error: pathspec 'gl-mod/bootstrap' did not match any file(s) known to git
logs/tools/libtool/compile.txt-bootstrap: running: git submodule update -- gl-mod/bootstrap
logs/tools/libtool/compile.txt-error: pathspec 'gl-mod/bootstrap' did not match any file(s) known to git
logs/tools/libtool/compile.txt-bootstrap: running: /Volumes/OpenWrt/openwrt/staging_dir/host/share/gnulib/gnulib-tool --no-changelog --avoid=dummy --libtool --macro-prefix=GL --with-tests --tests-base=gnulib-tests --aux-dir=build-aux --m4-base=m4 --local-dir=gl --local-dir=gl-mod/bootstrap --symlink --import announce-gen bootstrap...
logs/tools/libtool/compile.txt-gnulib-tool: warning: module bootstrap doesn't exist
logs/tools/libtool/compile.txt-gnulib-tool: warning: module extract-trace doesn't exist
logs/tools/libtool/compile.txt-gnulib-tool: warning: module inline-source doesn't exist
```
</details>

mpratt14 commented:
hmm....maybe libxml needs `-DLIBXML_WITH_ICONV` to be variable using `CONFIG_BUILD_NLS`

libtool didn't actually error out

but yeah that's unrelated. instead of building *literally* everything can you have CI do a standard snapshot style build?

mpratt14 commented:
this might fix libxml

https://github.com/mpratt14/openwrt/commit/cb0f2976df38e03568f54053aac34297216036ad

ldir-EDB0 commented:
Just to say the latest 'don't replace binaries with symlinks' commit has solved the problem for me.

ldir-EDB0 commented:
I think it should also leave links that point to staging_dir/host/bin alone.  Example:
From a make dirclean, make prereq puts symbolic links for staging_dir_host/bin/install to /usr/local/bin/ginstall
then make -j4:
ginstall gets built and installed as a binary in staging_dir/host/bin/ginstall and also a symlink of 'install' pointing to 'ginstall':
lrwxr-xr-x  1 kevin  staff  8 29 May 07:13 install -> ginstall
-rwxr-xr-x  1 kevin  staff  206840 29 May 07:13 ginstall

Anything that make prereq run will cause prereq to overwrite the 'install>ginstall' symbolic link with a link back to '/usr/local/bin/ginstall'

I wish I understood how to trigger prereq

ynezz commented:
> I wish I understood how to trigger prereq

`rm staging_dir/host/.prereq-build`

mcprat committed:
prereq-build: replace relative symlinks only if broken

Some programs installed to staging_dir/host/bin
also install some symlinks to itself
for an alternative name.

Some of those new symlinks are overwriting
symlinks that were installed by prereq stage.

If prereq stage were to somehow be run again,
it should not be overwriting symlinks
that point to programs that are already built.

To filter that out, catch all symlinks
after first catching all symlinks
that have an absolute target
after all other cases in the case statement,
make sure it is not broken, and if so exit successfully.

Suggested-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

ldir-EDB0 commented:
> > I wish I understood how to trigger prereq
> 
> `rm staging_dir/host/.prereq-build`

Yes, ok I can brute force it that way.  I'm curious what prompts a 'normal' check.

Anyway, with the brute force done, on my system there are 2 links that get replaced after a build, they are:

install & xargs.  The xargs one is very strange because that's a binary that gets built.  install is a symlink to ginstall in same dir.

mpratt14 commented:
> I'm curious what prompts a 'normal' check.

should be only controlled by that stampfile + include/prereq.mk + include/prereq-build.mk ... maybe another Makefile's timestamp not sure...

but running `make -j4` twice definitely should not be updating links the second time...

mpratt14 commented:
> I think it should also leave links that point to staging_dir/host/bin alone

great idea!

I hope the last commit here handles that cleanly enough, assuming all those links are relative targets only. And if not, I think they should be turned into relative links. Not sure how else to do it...

ldir-EDB0 commented:
the xargs binary that gets built is still being updated & replaced with a symbolic link.

-rwxr-xr-x  1 kevin  staff   112K 29 May 15:57 xargs
turns into
lrwxr-xr-x  1 kevin  staff    21B 29 May 19:16 xargs -> /usr/local/bin/gxargs

at least builds still work

mpratt14 commented:
ah yes, xargs doesn't have a stampfile because it's coreutils...

mpratt14 commented:
what if we did actually make stampfiles for all the coreutils tools that get installed?

mpratt14 commented:
wait its actually not from coreutils...

how does xargs get installed?

mcprat committed:
tools/coreutils: rename list of installed programs

Rename the list of programs installed by coreutils
to PKG_PROGRAMS, which will create a stampfile for each
through a new feature in host-build.mk.

Also, cleanup a bit to save lines
regarding the usage of this list.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
tools/findutils: define list of installed programs

Define the variable PKG_PROGRAMS for the list
of programs installed by findutils,
which will create a stampfile for each
through a new feature in host-build.mk.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
host-build: add support for a stampfile per installed binary

Some individual build items install a group of programs
instead of a program matching the name of the build item.

Add support for installing stampfiles for each of the
programs installed by that build item,
which will allow more control and awareness
of what is installed by the rest of the build system,
if, for example, prereq symlink checks are looking
for the same program which is built already.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
this should now also handle `xargs` being replaced with a symlink (and also `find` had the same situation I believe)

mpratt14 commented:
does the stampfile method look ok?

although I just realized I should make sure it rebuilds if the extra stamps are missing...

ldir-EDB0 commented:
It all works for me and I think this has dragged on long enough :-)

ynezz commented:
Thanks a lot, merged in 14a85d92
